src/changelog/ticketChangelog.ts

```diff
@@ -0,0 +1,116 @@
+/**
+ * Implements methods for working with ticket scopes, with the ability to search
+ * for common scopes both within the project and within the tags
+ */
+
+import * as E from "fp-ts/Either";
+import * as O from "fp-ts/Option";
+import * as RA from "fp-ts/ReadonlyArray";
+import { pipe } from "fp-ts/lib/function";
+import { GenericTicket } from "../types";
+import { popFromRegex, isSameScope } from "../utils/validator";
+import { Scope, projectToScope, tagToScope } from "./types";
+
+export declare function warn(message: string): void;
+export declare function markdown(message: string): void;
+export declare function schedule<T>(asyncFunction: Promise<T>): void;
+
+// pattern used to recognize a scope label
+const regex = /(changelog-scope:|epic-)(.*)/m;
+
+/**
+ *Try to detect the {@link Scope} of the ticket by projectId
+ */
+export const getProjectScope = (
+  ticket: GenericTicket
+): E.Either<Error, Scope> =>
+  pipe(
+    O.fromNullable(projectToScope[ticket.projectId]),
+    E.fromOption(
+      () =>
+        new Error(
+          `I can't find a scope related to the project ${ticket.projectId}`
+        )
+    )
+  );
+
+/**
+ * Try to detect the {@link Scope} of the ticket by tags associated
+ */
+export const getTagsScope = (ticket: GenericTicket): E.Either<Error, Scope> => {
+  const possibleScopes = pipe(
+    ticket.tags,
+    RA.map((l) => popFromRegex(l, regex)),
+    RA.filter(O.isSome)
+  );
+
+  if (possibleScopes.length === 0) {
```

Comment:
Avoid the use of imperative programming, you can rewrite this function composing your utilities with `fp-ts` functions such as Array's `size`, `Boolean.fold` etc.

Reply:
could this be correct?

```
return pipe(
    possibleScopes,
    RA.isEmpty,
    fold(
      () =>
        pipe(
          RA.size(possibleScopes) > 1,
          fold(
            () => getFirstScope,
            () =>
              E.left(
                new Error(
                  `Multiple labels match the expression \`${scopeLabelRegex}\` for the ticket [#${ticket.id}].\n
             It is not possible to assign a single scope to this pull request!`
                )
              )
          )
        ),
      () =>
        E.left(
          new Error(
            `No labels match the expression \`${scopeLabelRegex}\` for the ticket [#${ticket.id}].\n
       It is not possible to assign a scope to this pull request!`
          )
        )
    )
  );
```